packages/api: Add version information to prometheus metrics

hjpotter92 opened the pull request.

hjpotter92 committed:
packages/api: Add version information to prometheus metrics

hjpotter92 committed:
docker.yaml: Combine docker image generation to single CI workflow

victorges commented on .github/workflows/api.yaml:
```diff
@@ -1,88 +0,0 @@
-
-      - name: docker build & push
-        env:
-          DOCKER_TAGS: ${{ steps.tags.outputs.tags }}
-          VERSION: ${{ steps.tags.outputs.version }}
```
I wonder if we should remove the scripts/logic in the `package.json` to use these vars. If we're not using it anymore, might be preferred.

hjpotter92 replied:
cleaned up the scripts from `package.json` entirely.

victorges commented on .github/workflows/docker.yaml:
```diff
@@ -0,0 +1,98 @@
+      fail-fast: false
+      matrix:
+        build:
+          - path: packages/www
+            image: livepeerci/www
+            package: www
+          - path: packages/api
+            image: livepeerci/api
+            package: api
```
Oh nice!

victorges commented on packages/api/package.json:
```diff
@@ -15,7 +15,7 @@
     "docker:push": "for TAG in ${DOCKER_TAGS:-dev}; do docker push livepeerci/api:${TAG}; done",
     "docker": "run-s docker:build docker:push",
-    "postgres:start": "docker run --rm -d --name postgres -e POSTGRES_HOST_AUTH_METHOD=trust -p 5432:5432 -v \"$(pwd)/data/postgres:/var/lib/postgresql/data\" postgres",
+    "postgres:start": "docker run --rm -d --name postgres -e POSTGRES_HOST_AUTH_METHOD=trust -p 5432:5432 -v \"$(pwd)/data/postgres:/var/lib/postgresql/data\" postgres:14-alpine",
```
I think the tests on the API can be pretty flaky depending on these versions. Did you check if they (`yarn test`) work on the local dev env if running this postgres version?

hjpotter92 replied:
postgres (without tags) starts pgsql 15. the sql scripts in project are incompatible w/ 15, but work fine with 14. so i left the tag as 14. tests seem to pass anyway so not sure 🤷🏾

victorges commented on packages/api/src/index.ts:
```diff
@@ -109,6 +120,20 @@ export default async function makeApp(params: CliArgs) {
+        nodeversion: process.version,
+        arch: process.arch,
+        os: process.platform,
+        apiversion: getGitHash(),
```
I wonder if this should be just a global field in all metrics instead, like configured on the Prometheus client. Then we could also compare any metric based on the app version. It should not increase the cardinality, but not sure if it'd increase storage use significantly (victoria compresses the hell out of it tho, maybe not).

Also, I'd name this just `version`, as `www` doesn't really exist in the API server context. WDYt?

victorges commented on packages/www/Dockerfile:
```diff
@@ -1,12 +1,13 @@
-FROM node:14
+WORKDIR	/app
+
+COPY	package.json yarn.lock lerna.json	./
+
+COPY	.	.
+
```
Not a change here, but shouldn't the `yarn install` below go between these 2 copies? Otherwise there's no point in copying the `yarn` files first AFAIK

hjpotter92 replied:
that's what i thought too. but wasn't sure how lerna packages work here...

victorges reviewed (approved):
LGTM!

thomshutt commented on packages/api/src/index.ts:
```diff
@@ -109,6 +120,20 @@ export default async function makeApp(params: CliArgs) {
 
   if (listen) {
+    let version = new Gauge({
+      name: "livepeer_api_versions",
```
I think from Eli's comment in https://discord.com/channels/423160867534929930/993331529495351357/1060584233086759024 that we're going to standardise on a metric called "version" with "version" and "app" as labels

iameli replied:
Yeah, that's probably a good idea, but go-livepeer already uses `livepeerversion` so eh.

hjpotter92 committed:
metrics: Rename `apiversion` -> `version`

 - Add a metric label `app` for identification.
 - Drop all `docker:` scripts from package.json

iameli commented on packages/www/Dockerfile:
```diff
@@ -1,12 +1,13 @@
+
+RUN	yarn install --frozen-lockfile
+
+COPY	.	.
```
This will dump a developer's `node_modules` directory into any locally-built Docker images. Need a `.dockerignore` if we're going this route

victorges replied:
AFAICT nothing changed in this logic apart from changing `ADD` to `COPY`

hjpotter92 replied:
i added a .dockerignore to keep docker build context small when dev builds images locally

iameli commented on packages/www/package.json:
```diff
@@ -11,8 +11,6 @@
     "postbuild": "next-sitemap",
     "type-check": "tsc --pretty --noEmit",
     "docs-positions": "node ./generate-docs-positions.js",
-    "docker:build": "docker build $(printf ' -t livepeerci/www:%s' ${DOCKER_TAGS:-master}) -f Dockerfile ../..",
```
You can get rid of push maybe, but I've used this command to build the image locally in the past

hjpotter92 committed:
docker: Fix docker build issues
